core/frontend/src/QuantityFormatter.ts

```diff
@@ -83,382 +66,579 @@ const unitData: UnitDefinition[] = [
+  usSurvey?: FormatProps;
+}
+
+/** Represents both standard and custom quantity types
+ * @alpha
+ */
+export type QuantityTypeArg = QuantityType | string;
+
+/** Interface that defines the functions required to be implemented to provide custom formatting and parsing of a custom quantity type.
+ * @alpha
+ */
+export interface FormatterParserSpecsProvider {
+  quantityType: QuantityTypeArg;
+  createFormatterSpec: (unitSystem: UnitSystemKey) => Promise<FormatterSpec>;
+  createParserSpec: (unitSystem: UnitSystemKey) => Promise<ParserSpec>;
+}
+
+// ====================================== END OF DEFAULT DATA ======================================
+
+/** Mapping of FormatterSpecs by QuantityType. A FormatterSpec contains unit conversion factors for all composite units to go from the "persistence unit" to the
+ * display format without requiring an async unit look up. See method `getPersistenceUnitByQuantityType` for default persistence units.
+ * @internal
+ */
+export type FormatterSpecByQuantityType = Map<QuantityTypeKey, FormatterSpec>;
+
+/** Mapping of FormatterSpecs by QuantityType. A FormatterSpec contains unit conversion factors for all composite units to go from the "persistence unit" to the
+ * display format without requiring an async unit look up. See method `getPersistenceUnitByQuantityType` for default persistence units.
+ * @internal
+ */
+export type FormatPropsByUnitSystem = Map<UnitSystemKey, FormatProps>;
 
 /** Formats quantity values into strings.
  * @alpha
  */
 export class QuantityFormatter implements UnitsProvider {
-  protected _activeSystemIsImperial = true;
+  protected _activeUnitSystem: UnitSystemKey = "imperial";
   protected _formatSpecsByKoq = new Map<string, FormatterSpec[]>();
-  protected _imperialFormatsByType = new Map<QuantityType, Format>();
-  protected _metricFormatsByType = new Map<QuantityType, Format>();
-  protected _imperialFormatSpecsByType = new Map<QuantityType, FormatterSpec>();
-  protected _metricFormatSpecsByType = new Map<QuantityType, FormatterSpec>();
-  protected _imperialParserSpecsByType = new Map<QuantityType, ParserSpec>();
-  protected _metricUnitParserSpecsByType = new Map<QuantityType, ParserSpec>();
-  protected _overrideFormatDataByType = new Map<QuantityType, OverrideFormatEntry>();
+  protected _activeFormatSpecsByType = new Map<QuantityTypeKey, FormatterSpec>();
+  protected _activeParserSpecsByType = new Map<QuantityTypeKey, ParserSpec>();
+  protected _overrideFormatPropsByQuantityType = new Map<QuantityTypeKey, OverrideFormatEntry>();
+  protected _formatSpecProviders = new Array<FormatterParserSpecsProvider>();
+
+  /** Called after the active unit system is changed.
+   * The useImperial argument should not be relied on now that multiple systems are supported. The system parameter
+   * will contain the system name being activated.
+   */
+  public readonly onActiveUnitSystemChanged = new BeUiEvent<{ useImperial?: boolean, system?: string }>();
```

Comment:
Is there a reason why the arguments of the event are optional? The `useImperial` use to be required in onActiveUnitSystemChanged. (Same comment for onQuantityFormatsChanged.)

Reply:
As the comment states, the user should now use the 'system' argument instead. I kept the useImperial argument so I would not break any existing users, but wouldn't removing "?" from system be a breaking change?.

Reply:
Might be cleaner to deprecate old event and add new event with desired args.

Reply:
Good idea! That is what I will do.